cirq/google/arg_func_langs.py

```diff
@@ -12,9 +12,211 @@
+# Clients should use the least flexible language they can, to make it easier
+# to gradually roll out new capabilities to clients and servers.
+LANGUAGE_ORDER = [
+    '',
+    'linear',
+]
+
+
+def _max_lang(langs: Iterable[str]) -> str:
+    i = max((LANGUAGE_ORDER.index(e) for e in langs), default=0)
+    return LANGUAGE_ORDER[i]
+
+
+def _infer_function_language_from_circuit(value: v2.program_pb2.Circuit) -> str:
+    return _max_lang({
+        e for moment in value.moments for op in moment.operations
+        for e in _function_languages_from_operation(op)
+    })
+
+
+def _infer_function_language_from_schedule(value: v2.program_pb2.Schedule
+                                          ) -> str:
+    return _max_lang({
+        e for op in value.scheduled_operations
+        for e in _function_languages_from_operation(op.operation)
+    })
+
+
+def _function_languages_from_operation(value: v2.program_pb2.Operation
+                                      ) -> Iterator[str]:
+    for arg in value.args.values():
+        yield from _function_languages_from_arg(arg)
+
+
+def _function_languages_from_arg(arg_proto: v2.program_pb2.Arg
+                                ) -> Iterator[str]:
+
+    which = arg_proto.WhichOneof('arg')
+    if which == 'func':
+        if arg_proto.func.type in ['add', 'mul']:
+            yield 'linear'
+            for a in arg_proto.func.args:
+                yield from _function_languages_from_arg(a)
+
+
+def _arg_to_proto(value: ARG_LIKE,
+                  *,
+                  arg_function_language: Optional[str],
+                  out: Optional[v2.program_pb2.Arg] = None
+                 ) -> v2.program_pb2.Arg:
+    """Writes an argument value into an Arg proto.
+
+    Args:
+        value: The value to encode.
+        arg_function_language: The language to use when encoding functions. If
+            this is set to None, it will be set to the minimal language
+            necessary to support the features that were actually used.
+        out: The proto to write the result into. Defaults to a new instance.
+
+    Returns:
+        The proto that was written into as well as the `arg_function_language`
+        that was used.
+    """
+
+    if arg_function_language not in SUPPORTED_FUNCTIONS_FOR_LANGUAGE:
+        raise ValueError(f'Unrecognized arg_function_language: '
+                         f'{arg_function_language!r}')
+    supported = SUPPORTED_FUNCTIONS_FOR_LANGUAGE[arg_function_language]
+
+    msg = v2.program_pb2.Arg() if out is None else out
+
+    def check_support(func_type: str) -> str:
+        if func_type not in supported:
+            lang = (repr(arg_function_language)
+                    if arg_function_language is not None else '[any]')
+            raise ValueError(f'Function type {func_type!r} not supported by '
+                             f'arg_function_language {lang}')
+        return func_type
+
+    if isinstance(value,
```

Comment:
> Would this miss things like `sympy.pi` ?

I added `sympy.NumberSymbol` which will handle pi by turning it into a float. It's okay if we miss some of these cases initially because when we serialize we don't have to remember if it was the symbol or the raw number.

> putting all of these if's elif's in a seperate function

This function is already the extracted function for holding all the if's and elif's :stuck_out_tongue: ! We'll refactor as needed as it grows.